archive_xlog.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH ARCHIVE_XLOG 1 "October 11, 2015"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
archive_xlog \- save WAL segment files for point-in-time recovery

.SH SYNOPSIS
.B archive_xlog
.RI [ options ]
.I xlogfile


.SH DESCRIPTION
The \fBarchive_xlog\fP script may be used as the \fIarchive_command\fP in
\fIpostgresql.conf\fP to copy and optionally compress completed WAL segment
files to a local or remote archive location.  It is not required to use this
script as your \fIarchive_command\fP when using \fBpitrery\fP for backup and
restore operations, but it does make it a bit easier to keep all of the
relevant configuration for storing and retrieving WAL segment files in the
fewest number of places possible, with the least risk of them accidentally
getting out of sync if the configuration is changed.


.SH OPTIONS
The following options are available:

.TP
.BI "\-C " conf
Use option settings from the specified configuration file instead of the
default \fIpitrery.conf\fP file.  This may be a full path to the file to use,
or the name of a configuration file in \fI/etc/pitrery\fP.

.TP
.BI "\-a " [[user@]host:]/dir
The directory on the (local or remote) host where WAL segment files will be
stored.  This will override the \fBARCHIVE_DIR\fP option from the
configuration file.  When the host name is omited, the archiving is local.

.TP
.B \-X
Do not compress WAL segment files.  This will override the
\fBARCHIVE_COMPRESS\fP option from the configuration file (forcing it to be
"no").

.TP
.B \-O
Do not overwrite the destination file if it exists.  This will force
the \fBARCHIVE_OVERWRITE\fP option from the configuration file to "no".

.TP
.B \-H
Check the md5 of the archived file to the md5 of the original WAL
file. While adds the overhead of computing the checksums of the source
and destination files, it is useful if the network or the remote
storage is not considered reliable. This will force the
\fBARCHIVE_CHECK\fP option from the configuration file to "yes". This
option does not apply on local archiving.

.TP
.B \-F
Force an immediate flush of the archived file to disk before returning
success. It may slow down the archiving process but ensure archives
are not corrupted in case of a power loss on the destination. This will force the
\fARCHIVE_FLUSH\fP option from the configuration file to "yes".

.TP
.BI "\-c " command
The command to use for compressing WAL segment files.  This will override
the \fBARCHIVE_COMPRESS_BIN\fP option from the configuration file.

.TP
.BI "\-s " suffix
The compressed file suffix to use (e.g., gz, bz2, xz).  This will override
the \fBARCHIVE_COMPRESS_SUFFIX\fP option from the configuration file.

.TP
.BI \-E
Encrypt the file using GnuPG.  The list of recipients must given using
the \fB-r\fP option.  When this option is used, compression is
disabled, as GunPG already compresses the output.  This will override
the \fBARCHIVE_ENCRYPT\fP option from the configuration file.

.TP
.BI "\-r " keys:...
List of public keys for GnuPG encryption, it is a colon separeted list
of USER-ID recognized by the \fB--recipient\fP option of \fBgpg\fP(1).
This will override the \fBGPG_ENCRYPT_KEYS\fP option from the
configuration file.

.TP
.B \-S
Send messages to the \fBsyslog\fP(3) instead of \fBstdio\fP(3).  This will
override the \fBSYSLOG\fP option from the configuration file (forcing it to
be "yes").  Messages sent to \fIstdout\fP will be logged with \fBLOG_INFO\fP
priority, while messages to \fIstderr\fP will be logged with \fBLOG_ERR\fP
priority.

.TP
.BI "\-f " facility
Specify the \fBsyslog\fP facility to use.  This will override the
\fBSYSLOG_FACILITY\fP option from the configuration file.  Defaults to use
\fBlocal0\fP if not otherwise explicitly set.  See \fBlogger\fP(1) for details
of the valid facility strings that can be used here.

.TP
.BI "\-t " ident
An identifier to prefix \fBsyslog\fP output with.  This will override the
\fBSYSLOG_IDENT\fP option from the configuration file.  Defaults to use the
string "postgres" if not otherwise explicitly set.

.TP
.BI "\-m " mode
Set the permissions of the archived file to this mode (e.g. \fBchmod\fP).  The value
must be in octal form.

.TP
.B \-T
Timestamp the log messages.  This will override the \fBLOG_TIMESTAMP\fP option
from the configuration file (forcing it to be "yes").

.TP
.B \-?
Output the brief help summarising the above options.

.TP
.I xlogfile
The path to the WAL segment to archive. It must be set to \fB%p\fP inside
\fIpostgresql.conf\fP when configuring archive_command.


.SH EXAMPLES
A minimal \fIpostgresql.conf\fP configuration for WAL archiving with
\fBarchive_xlog\fP might look something like:

.nh
.nf
  # If using PostgreSQL >= 9.0,
  # wal_level must be set to archive or higher.
  # Changing this requires a restart. If using PostgreSQL >= 9.6,
  # wal_level must be set to replica or higher
  wal_level = archive

  # If using PostgreSQL >= 8.3, archiving must be enabled.
  # Changing this requires a restart.
  archive_mode = on

  # The archive command using the defaults.
  # Changing this requires a reload.
  archive_command = 'archive_xlog %p'

  # The archive command with parameters.
  #archive_command = 'archive_xlog \-C /path/to/pitrery.conf %p'
  # or to search /etc/pitrery for the pitrery.conf configuration:
  #archive_command = 'archive_xlog \-C pitrery %p'
.fi
.hy

The \fIpitrery.conf\fP that will be used should be configured suitably before
these changes are made to the \fBpostgres\fP configuration and the postmaster
is restarted or reloaded.  Once the above configuration is active, the
\fBarchive_xlog\fP script may be called at any time.  See \fBpitrery\fP(1)
for details of the configuration options available there.


.SH FILES
.TP
.I /etc/pitrery/pitrery.conf
The default \fBarchive_xlog\fP configuration file if not explicitly specified.

.TP
.I /etc/postgresql/**/postgresql.conf $PGDATA/postgresql.conf
The \fBpostgres\fP(1) configuration file where \fIarchive_command\fP (and
other WAL archiving configuration) must be set to use \fBpitrery\fP (or any
other point-in-time recovery methods).


.SH SEE ALSO
.BR restore_xlog (1),
.BR pitrery (1).


.SH AUTHOR
.B archive_xlog
was written by Nicolas Thauvin <nicolas.thauvin@dalibo.com>.
This man page was written by Ron <ron@debian.org>.



CHANGELOG
pitrery 2.1
===========

  * Restrict permissions of backup directories to the user that created them

  * Add options to the check action to verify backup policy and archived WAL
    files

  * Add RSYNC_WHOLEFILE and RSYNC_BWLIMIT configuration options to
    tweak the behaviour of rsync when performing base backups.



pitrery 2.0
===========

  * Regroup all action into the main script

  * Change the sed commands in the Makefile to have the script work
    without modifications

  * Command line switches and options to specify if the backup is
    local, the user, host and target directory, are now merged into a
    SSH style syntax: `[[user@]host:]/path`. Not providing a host
    tells pitrery the backups are local.

  * Rename the default configuration file from `pitr.conf` to `pitrery.conf`

  * Comment out all parameters in the configuration file so that they
    do not interfere

  * Remove the BACKUP_LABEL option, subdirectory and -l switch. They
    were not used.

  * Support for PostgreSQL 10


pitrery 1.13
============

  * Fix tablespaces links being overwritten by postmaster on 9.5 due
    to tablespace_map. This conflicted with our tablespace relocation

  * Fix the purge of WAL file that would remove all archives if the
    stop file is missing

  * Fix backups failing due to incorrect parsing of the stop time
    when iso8601 timestamps are used


pitrery 1.12
============

  * New -m option for archive_xlog to chmod the archived file to the
    specified octal permission bits

  * Fix the restore so that the backup_label is restore whan used with
    PostgreSQL 9.6

  * Remove support of deprecated compression parameters, renamed in 1.9


pitrery 1.11
============

  * New configure action to create a configuration file from the
    command line.

  * New check action to test configuration files.

  * Do not backup replication slots state files and produce an SQL
    file to recreate them after restore.

  * Add a dry run option (-N) for purge.

  * Improve the quoting and ensure unusual characters are handled
    properly.

  * Add manpages for the user facing commands, pitrery, archive_xlog
    and restore_xlog.

  * Relax restrictions on the target date format on restore. It was
    meant to have a working format common to all free unices
    platforms, but the user can just adapt the command to its
    system. So now one can do run: pitrery restore -d '1 hour ago', or
    anything the local date command can parse.

  * New ARCHIVE_OVERWRITE configuration parameter along with -O option
    (do not overwrite) for archive_xlog to check if the target files
    exists before archiving, when set to "no" or -O is on the command
    line.

  * New ARCHIVE_CHECK (-H) and ARCHIVE_FLUSH (-F) for archive_xlog. The
    first one check md5 sum of the archived WAL file, the second force
    a flush to disk.

  * Make use of the new backup API of PostgreSQL 9.6

  * Optionally name backups dir with ISO 8601 timestamps, with config
    option USE_ISO8601_TIMESTAMPS and command line switch -t for the
    backup action.

  * French translation of the documentation.

  * Bugfixes and better handling of unusual characters in paths


pitrery 1.10
============

  * Fix issue in restore_xlog not using SSH user and host found in the
    configuration file, thus preventing restore to work out of the
    box.

  * Ensure archive_xlog create the target directory on remote hosts

  * Only purge WAL files when there are some WAL files to remove


pitrery 1.9
===========

  * Do not exit without error when the server is in recovery. This is
    not a consistent behaviour.

  * Improve the speed of the purge of archives WAL files by stopping
    the purge when the limit is reached.

  * Add the -T switch along with the LOG_TIMESTAMP option to prefix
    messages of backup, restore and purge with the date and time.

  * Rename the compression configuration parameters for WAL files to
    avoid confusion :
    * `COMPRESS_BIN` -> `ARCHIVE_COMPRESS_BIN`
    * `COMPRESS_SUFFIX` -> `ARCHIVE_COMPRESS_SUFFIX`
    * `UNCOMPRESS_BIN` -> `ARCHIVE_UNCOMPRESS_BIN`

  * Add the -c and -s switch to archive_xlog and restore_xlog which
    allow to give the (un)compression command and compressed file suffix
    respectively.

  * No longer use command line switch to configure restore_xlog for
    restore_command in restore action. Give a configuration file or
    the full command to use. Thus only -r and -C (changed to give the
    configuration file) remains for archived WAL options.

  * Add the -l switch to pitrery. It lists the configuration files
    inside the default configuration directory.

  * Include all recovery parameters commented out in the recovery.conf
    generated at restore.


pitrery 1.8
===========

  * Exclude temporary locations from backup

  * Support pax to create hardlinks, for better portability. GNU cp
    stays the default, the configuration goes to the Makfile.

  * Add a -V switch to the pitrery wrapper script to show the version.

  * Add an option (-R) to overwrite target directory on restore.


pitrery 1.7
===========

  * Show the storage method and compression suffix when listing
    backups in verbose mode.

  * Allow configuration of compression tools used when storing backup
    contents using tar.

  * Remove the `pitr_mgr` symlink.

  * Have `archive_xlog` create the local target directory so that we don't
    end up archiving on the same path, when the user forget to create it.

  * Run backup hook command even if the backup fails. The exit code of the
    backup is available to the hook in the PITRERY_EXIT_CODE env variable.

  * Bugfix on sed -E vs sed -r, use the later for older GNU sed.


pitrery 1.6
===========

  * Exit without error when backup is run on a hot standby
    server. Issue a warning message when run interactively, so an
    active cron job does not send any mail.

  * Bugfixes and improvements on documentation


pitrery 1.5
===========


archive_xlog.bash
#!@BASH@
#
# Copyright 2011-2017 Nicolas Thauvin. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 
#  1. Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
#  2. Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

# Apply an extra level of shell quoting to each of the arguments passed.
# This is necessary for remote-side arguments of ssh (including commands that
# are executed by the remote shell and remote paths for scp and rsync via ssh)
# since they will strip an extra level of quoting off on the remote side.
# This makes it safe for them to include spaces or other special characters
# which should not be interpreted or cause word-splitting on the remote side.
qw() {
    while (( $# > 1 )); do
	printf "%q " "$1"
	shift
    done
    (( $# == 1 )) && printf "%q" "$1"
}

# Message functions
now() {
    [ "$LOG_TIMESTAMP" = "yes" ] && echo "$(date "+%F %T %Z ")"
}

error() {
    echo "$(now)ERROR: $1" 1>&2
    [ -n "$tmpfile" ] && rm -f -- "$tmpfile"
    exit ${2:-1}
}

warn() {
    echo "$(now)WARNING: $*" 1>&2
}

# Script help
usage() {
    echo "usage: `basename $0` [options] XLOGFILE"
    echo "options:"
    echo "    -L             do local archiving"
    echo "    -C conf        configuration file"
    echo "    -u username    username for SSH login"
    echo "    -h hostname    hostname for SSH login"
    echo "    -d dir         target directory"
    echo "    -X             do not compress"
    echo "    -O             do not overwrite the destination file"
    echo "    -H             check the hash of the destination file (remote only)"
    echo "    -F             flush the destination file to disk"
    echo "    -c command     compression command"
    echo "    -s suffix      compressed file suffix (ex: gz)"
    echo "    -S             send messages to syslog"
    echo "    -f facility    syslog facility"
    echo "    -t ident       syslog ident"
    echo "    -m mode        destination file permission mode in octal (e.g. chmod)"
    echo
    echo "    -T             Timestamp log messages"
    echo "    -?             print help"
    echo
    exit $1
}

check_md5() {
    [ "$ARCHIVE_CHECK" != "yes" ] && return 0

    local ARCHIVE_MD5=$1

    LOCAL_MD5=$(md5sum "$xlog")

    if [ "${LOCAL_MD5%% *}" != "${ARCHIVE_MD5%% *}" ]; then
        error "md5 mismatch between local and remote file" 4
    fi

    return 0
}

# Configuration defaults
CONFIG=pitr.conf
CONFIG_DIR=@SYSCONFDIR@
ARCHIVE_DIR=/var/lib/pgsql/archived_xlog
ARCHIVE_LOCAL="no"
SYSLOG="no"
ARCHIVE_COMPRESS="yes"
ARCHIVE_COMPRESS_BIN="gzip -f -4"
ARCHIVE_COMPRESS_SUFFIX="gz"
ARCHIVE_OVERWRITE="yes"
ARCHIVE_CHECK="no"
ARCHIVE_FLUSH="no"
ARCHIVE_FILE_CHMOD=""

# Command line options
while getopts "LC:u:d:h:XOHFc:s:Sf:t:m:T?"  opt; do
    case $opt in
        L) CLI_ARCHIVE_LOCAL="yes";;
        C) CONFIG=$OPTARG;;
        u) CLI_ARCHIVE_USER=$OPTARG;;
        h) CLI_ARCHIVE_HOST=$OPTARG;;
        d) CLI_ARCHIVE_DIR=$OPTARG;;
        X) CLI_ARCHIVE_COMPRESS="no";;
        O) CLI_ARCHIVE_OVERWRITE="no";;
        H) CLI_ARCHIVE_CHECK="yes";;
        F) CLI_ARCHIVE_FLUSH="yes";;
        c) CLI_ARCHIVE_COMPRESS_BIN=$OPTARG;;
        s) CLI_ARCHIVE_COMPRESS_SUFFIX=$OPTARG;;
        S) CLI_SYSLOG="yes";;
        f) CLI_SYSLOG_FACILITY=$OPTARG;;
        t) CLI_SYSLOG_IDENT=$OPTARG;;
        m) CLI_ARCHIVE_FILE_CHMOD=$OPTARG;;
        T) CLI_LOG_TIMESTAMP="yes";;
        "?") usage 1;;
        *) error "Unknown error while processing options";;
    esac
done	

# Check if the config option is a path or just a name in the
# configuration directory.  Prepend the configuration directory and
# .conf when needed.
if [[ $CONFIG != */* ]]; then
    CONFIG="$CONFIG_DIR/$(basename -- "$CONFIG" .conf).conf"
fi

# Load configuration file
if [ -f "$CONFIG" ]; then
    . "$CONFIG"
fi

# Overwrite configuration with cli options
[ -n "$CLI_ARCHIVE_LOCAL" ] && ARCHIVE_LOCAL=$CLI_ARCHIVE_LOCAL
[ -n "$CLI_ARCHIVE_USER" ] && ARCHIVE_USER=$CLI_ARCHIVE_USER
[ -n "$CLI_ARCHIVE_HOST" ] && ARCHIVE_HOST=$CLI_ARCHIVE_HOST
[ -n "$CLI_ARCHIVE_DIR" ] && ARCHIVE_DIR=$CLI_ARCHIVE_DIR
[ -n "$CLI_ARCHIVE_COMPRESS" ] && ARCHIVE_COMPRESS=$CLI_ARCHIVE_COMPRESS
[ -n "$CLI_ARCHIVE_COMPRESS_BIN" ] && ARCHIVE_COMPRESS_BIN=$CLI_ARCHIVE_COMPRESS_BIN
[ -n "$CLI_ARCHIVE_COMPRESS_SUFFIX" ] && ARCHIVE_COMPRESS_SUFFIX=$CLI_ARCHIVE_COMPRESS_SUFFIX
[ -n "$CLI_ARCHIVE_OVERWRITE" ] && ARCHIVE_OVERWRITE=$CLI_ARCHIVE_OVERWRITE
[ -n "$CLI_ARCHIVE_CHECK" ] && ARCHIVE_CHECK=$CLI_ARCHIVE_CHECK
[ -n "$CLI_ARCHIVE_FLUSH" ] && ARCHIVE_FLUSH=$CLI_ARCHIVE_FLUSH
[ -n "$CLI_SYSLOG" ] && SYSLOG=$CLI_SYSLOG
[ -n "$CLI_SYSLOG_FACILITY" ] && SYSLOG_FACILITY=$CLI_SYSLOG_FACILITY
[ -n "$CLI_SYSLOG_IDENT" ] && SYSLOG_IDENT=$CLI_SYSLOG_IDENT
[ -n "$CLI_ARCHIVE_FILE_CHMOD" ] && ARCHIVE_FILE_CHMOD=$CLI_ARCHIVE_FILE_CHMOD
[ -n "$CLI_LOG_TIMESTAMP" ] && LOG_TIMESTAMP=$CLI_LOG_TIMESTAMP

# Redirect output to syslog if configured
if [ "$SYSLOG" = "yes" ]; then
    SYSLOG_FACILITY=${SYSLOG_FACILITY:-local0}
    SYSLOG_IDENT=${SYSLOG_IDENT:-postgres}

    exec 1> >(logger -t "$SYSLOG_IDENT" -p "${SYSLOG_FACILITY}.info")
    exec 2> >(logger -t "$SYSLOG_IDENT" -p "${SYSLOG_FACILITY}.err")
fi

# The first argument must be a WAL file
xlog=${@:$OPTIND:1}
if [ -z "$xlog" ]; then
    error "missing xlog filename to archive. Please consider modifying archive_command, eg add %p"
fi

# Sanity check. We need at least to know if we want to perform a local
# copy or have a hostname for an SSH copy
if [ "$ARCHIVE_LOCAL" != "yes" ] && [ -z "$ARCHIVE_HOST" ]; then
    error "Not enough information to archive the segment"
fi

if [ "$ARCHIVE_LOCAL" = "yes" ] && [ -n "$ARCHIVE_HOST" ]; then
    error "ARCHIVE_LOCAL and ARCHIVE_HOST are set, it can't be both"
fi

# Check if the source file exists
if [ ! -r "$xlog" ]; then
    error "Input file '$xlog' does not exist or is not readable"
fi

# Set dd flush mode if needed
if [ "$ARCHIVE_FLUSH" = "yes" ]; then
    ARCHIVE_FLUSH='conv=fsync'
else
    ARCHIVE_FLUSH=''
fi

check_local_dest_exists()
{
    [ $ARCHIVE_OVERWRITE = "yes" ] && return 0

    [ ! -e "$1" ] || error "$1 already exists, refusing to overwrite it."
}

# Copy the wal locally
if [ "$ARCHIVE_LOCAL" = "yes" ]; then
    dd_rc=0
    mkdir -p -- "$ARCHIVE_DIR" 1>&2 ||
        error "Unable to create target directory '$ARCHIVE_DIR'" $?

    if [ "$ARCHIVE_COMPRESS" = "yes" ]; then
        dest_path=$ARCHIVE_DIR/$(basename -- "$xlog").$ARCHIVE_COMPRESS_SUFFIX
        check_local_dest_exists "$dest_path"

        $ARCHIVE_COMPRESS_BIN -c < "$xlog" | dd $ARCHIVE_FLUSH of="$dest_path" 2>/dev/null
        rc=( ${PIPESTATUS[@]} )
        x_rc=${rc[0]}
        dd_rc=${rc[1]}
        if [ $x_rc != 0 ]; then
            rm -f -- "$dest_path"
            error "Compressing $xlog to $dest_path failed"
        fi
    else
        dest_path=$ARCHIVE_DIR/$(basename -- "$xlog")
        check_local_dest_exists "$dest_path"

        dd $ARCHIVE_FLUSH if="$xlog" of="$dest_path" 2>/dev/null
        dd_rc=$?
    fi

    if [ $dd_rc != 0 ] ; then
        rm -f -- "$dest_path"
        error "Copying $xlog to $dest_path failed"
    fi

    # Chmod if required
    if [ -n "$ARCHIVE_FILE_CHMOD" ]; then
        echo "$ARCHIVE_FILE_CHMOD" | grep -qE '^[0-7]{3,4}$'
        if [ $? = 0 ]; then
            if ! chmod $ARCHIVE_FILE_CHMOD "$dest_path"; then
                warn "Could not change mode of $dest_path to $ARCHIVE_FILE_CHMOD"
            fi
        else
            warn "ARCHIVE_FILE_CHMOD is not in octal form, mode not changed"
        fi
    fi

else
    # Compress and copy with rsync
    echo $ARCHIVE_HOST | grep -qi '^[0123456789abcdef:]*:[0123456789abcdef:]*$' && ARCHIVE_HOST="[${ARCHIVE_HOST}]" # Dummy test for IPv6

    dest_host=${ARCHIVE_USER:+$ARCHIVE_USER@}${ARCHIVE_HOST}

    dest_file="$ARCHIVE_DIR/$(basename -- "$xlog")"
    tmp_file=""
    src_file="$xlog"

    # Depending on the options, we may check different things on the
    # remote host. To avoid many connections, we build a command to be
    # run one time on the remote host.

    # Create remote folder if needed. Return 2 on error.
    REMOTE_CMD="mkdir -p -- $(qw "$ARCHIVE_DIR") || exit 2"

    # Compress the file to a temporary location
    if [ "$ARCHIVE_COMPRESS" = "yes" ]; then
        dest_file=$ARCHIVE_DIR/$(basename -- "$xlog").$ARCHIVE_COMPRESS_SUFFIX
        tmp_file=$(mktemp -t pitr_wal.XXXXXXXXXX) ||
            error "Failed to create temporary file for compressed WAL" $?

        src_file="$tmp_file"

        # We take no risk, pipe the content to the compression program
        # and save output elsewhere: the compression program never
        # touches the input file
        $ARCHIVE_COMPRESS_BIN -c < "$xlog" > "$tmp_file" ||
            error "Compressing $xlog to $tmp_file failed" $?
    fi

    # Check if the file exists on the remote host. If the file exists
    # on the remote host and we do not overwrite it, we just get its
    # md5 sum and exit. Later the check of the md5 will exit on error
    # if the sum a different
    if [ "$ARCHIVE_OVERWRITE" != "yes"  ] && [ "$ARCHIVE_CHECK" = "yes" ]; then
        REMOTE_CMD="$REMOTE_CMD; [ -e $(qw "$dest_file") ] && echo \$(md5sum -- $(qw "$dest_file")) && exit 0"
    elif [ "$ARCHIVE_OVERWRITE" != "yes" ]; then
        REMOTE_CMD="$REMOTE_CMD; [ ! -e $(qw "$dest_file") ] || exit 3"
    fi

    # Copy the file with dd
    REMOTE_CMD="$REMOTE_CMD; dd $ARCHIVE_FLUSH of=$(qw "$dest_file") 2>/dev/null || exit 1"

    if [ "$ARCHIVE_CHECK" = "yes" ]; then
        REMOTE_CMD="$REMOTE_CMD; md5sum -- $(qw "$dest_file")"
    fi

    # Chmod if required
    if [ -n "$ARCHIVE_FILE_CHMOD" ]; then
        echo "$ARCHIVE_FILE_CHMOD" | grep -qE '^[0-7]{3,4}$'
        if [ $? = 0 ]; then
            REMOTE_CMD="$REMOTE_CMD; chmod $ARCHIVE_FILE_CHMOD $(qw "$dest_file") || exit 4"
        else
            warn "ARCHIVE_FILE_CHMOD is not in octal form, mode not changed"
        fi
    fi

    # Actually execute the remote commands
    remote_md5=$(dd if="$src_file" 2>/dev/null|ssh -- "$dest_host" "$REMOTE_CMD")
    rc=$?

    case $rc in
        0) ;;
        1) error "Unable to copy $xlog to ${ARCHIVE_HOST}:${ARCHIVE_DIR}" $rc;;
        2) error "Unable to create target directory" $rc;;
        3) error "'$dest_file' already exists on $dest_host, refusing to overwrite it" $rc;;
        4) warn "Could not change mode of $dest_path to $ARCHIVE_FILE_CHMOD";;
        255) error "SSH error on ${ARCHIVE_HOST}" $rc;;
        *) error "Unexpected return code while copying the file" 100
    esac

    if [ "$ARCHIVE_CHECK" = "yes" ]; then
        local_md5=$(md5sum -- "$src_file")

        if [ "${local_md5%% *}" != "${remote_md5%% *}" ]; then
            error "md5 mismatch between local and remote file" 4
        fi
    fi

    # Remove temp file if exists
    if [ -n "$tmp_file" ] && [ -s "$tmp_file" ]; then
        rm -- "$tmp_file" ||
            warn "Unable to remove temporary compressed file '$tmp_file'"
    fi

fi

exit 0
